storm-core/src/jvm/org/apache/storm/ProcessSimulator.java

```diff
@@ -0,0 +1,89 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.storm;
+import org.apache.storm.daemon.Shutdownable;
+
+import java.util.Collection;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ProcessSimulator {
+    private static Logger LOG = LoggerFactory.getLogger(ProcessSimulator.class);
+    protected static Object lock = new Object();
+    protected static ConcurrentHashMap<String, Shutdownable> processMap = new ConcurrentHashMap<String, Shutdownable>();
+
+    /**
+     * Register a process' handle
+     * 
+     * @param pid
+     * @param shutdownable
+     */
+    public static void registerProcess(String pid, Shutdownable shutdownable) {
+        processMap.put(pid, shutdownable);
+    }
+
+    /**
+     * Get a process' handle
+     * 
+     * @param pid
+     * @return
+     */
+    protected static Shutdownable getProcessHandle(String pid) {
+        return processMap.get(pid);
+    }
+
+    /**
+     * Get all process handles
+     * 
+     * @return
+     */
+    public static Collection<Shutdownable> getAllProcessHandles() {
+        return processMap.values();
+    }
+
+    /**
+     * Kill a process
+     * 
+     * @param pid
+     */
+    public static void killProcess(String pid) {
+        synchronized (lock) {
+            LOG.info("Begin killing process " + pid);
+            Shutdownable shutdownHandle = getProcessHandle(pid);
+            if (shutdownHandle != null) {
+                shutdownHandle.shutdown();
+            }
+            processMap.remove(pid);
```

Comment:
Here I would prefer to keep it unchanged since It would be possible that <pid, null> be wrongly put in the map by registerPorcess. Place the "remove" outside of the "if" would help and not hurt, also this translation conforms the original clj code logic.